src/xwayland/xwm/mod.rs

```diff
@@ -657,6 +666,57 @@ impl X11WM {
 
         surface.state.lock().unwrap().wl_surface = Some(wl_surface);
     }
+
+    /// Set the default cursor used by X clients
+    pub fn set_cursor(
+        &mut self,
+        pixels: &[u8],
+        size: Size<u16, Logical>,
+        hotspot: Point<u16, Logical>,
+    ) -> Result<(), ReplyOrIdError> {
+        let pixmap = PixmapWrapper::create_pixmap(&*self.conn, 32, self.screen.root, size.w, size.h)?;
+        let Some(render_format) = self
+            .conn
+            .render_query_pict_formats()?
+            .reply_unchecked()?
+            .unwrap_or_default()
+            .formats
+            .into_iter()
+            .filter(|f| f.depth == 32)
+            .map(|f| f.id)
+            .next()
+        else {
+            return Err(ReplyOrIdError::ConnectionError(ConnectionError::UnknownError)); // TODO proper error type
+        };
+        let picture = PictureWrapper::create_picture(
+            &*self.conn,
+            pixmap.pixmap(),
+            render_format,
+            &CreatePictureAux::new(),
+        )?;
+        let gc = GcontextWrapper::create_gc(&*self.conn, picture.picture(), &CreateGCAux::new())?;
+        self.conn.put_image(
+            ImageFormat::Z_PIXMAP,
+            picture.picture(),
+            gc.gcontext(),
+            size.w,
+            size.h,
+            0,
+            0,
+            0,
+            32,
+            pixels,
+        )?;
+        let cursor = self.conn.generate_id()?;
+        self.conn
+            .render_create_cursor(cursor, picture.picture(), hotspot.x, hotspot.y)?;
+        if let Some(old_cursor) = self.cursor.replace(cursor) {
+            let _ = self.conn.free_cursor(old_cursor);
+        };
+        self.conn
+            .change_window_attributes(self.screen.root, &ChangeWindowAttributesAux::new().cursor(cursor))?;
```

Comment:
AFAIR, X11 resources are usually reference counted internally in the X11 server. Thus, you should be able to `free_cursor` directly after `change_window_attributes()`. But I am not a 100% sure about this.

(And I am only guessing that `self.cursor` exists to only free the cursor when another one is needed.)

Reply:
Seems like it works! thanks :)